modules/snmp-dest/snmpdest.c

```diff
@@ -0,0 +1,899 @@
+      snmp_shutdown(s_snmp_name);
+    }
+  --snmp_dest_counter;
+
+  g_mutex_free(self->queue_mutex);
+  g_cond_free(self->writer_thread_wakeup_cond);
+
+  g_free(self->version);
+  g_free(self->host);
+  if (self->host_port)
+    g_string_free(self->host_port, TRUE);
+  while(self->snmp_objs)
+    {
+      g_free(self->snmp_objs->data);
+      self->snmp_objs = g_list_delete_link(self->snmp_objs, self->snmp_objs);
+    }
+  while(self->snmp_codes)
+    {
+      g_free(self->snmp_codes->data);
+      self->snmp_codes = g_list_delete_link(self->snmp_codes, self->snmp_codes);
+    }
+  while (self->snmp_templates)
+    {
+      log_template_unref(self->snmp_templates->data);
+      self->snmp_templates = g_list_delete_link(self->snmp_templates, self->snmp_templates);
+    }
+  g_free(self->snmp_objs);
+  if (self->trap_oid)
+    g_free(self->trap_oid);
+  if (self->trap_type)
+    g_free(self->trap_type);
+  if (self->trap_oid)
+    g_free(self->trap_value);
+  g_free(self->community);
+  if (self->engine_id)
+    g_free(self->engine_id);
+  if (self->auth_username)
+    g_free(self->auth_username);
+  g_free(self->auth_algorithm);
+  if (self->auth_password)
+    g_free(self->auth_password);
+  g_free(self->enc_algorithm);
+  if (self->enc_password)
+    g_free(self->enc_password);
+  g_free(self->transport);
+
+  log_template_options_destroy(&self->template_options);
+
+  log_dest_driver_free(d);
+}
+
+static void
+snmpdest_dd_queue(LogPipe *s, LogMessage *msg, const LogPathOptions *path_options)
+{
+  SNMPDestDriver *self = (SNMPDestDriver *)s;
+
+  g_mutex_lock(self->queue_mutex);
+
+  stats_counter_inc(self->processed_messages);
+  log_queue_push_tail(self->queue, msg, path_options);
+
+  g_mutex_unlock(self->queue_mutex);
+
+  g_cond_signal(self->writer_thread_wakeup_cond);
+}
+
+LogDriver *
+snmpdest_dd_new(GlobalConfig *cfg)
+{
+  SNMPDestDriver *self = g_new0(SNMPDestDriver, 1);
+
+  log_dest_driver_init_instance(&self->super, cfg);
+  self->super.super.super.init = snmpdest_dd_init;
+  self->super.super.super.deinit = snmpdest_dd_deinit;
+  self->super.super.super.queue = snmpdest_dd_queue;
+  self->super.super.super.free_fn = snmpdest_dd_free;
+
+  self->writer_thread_wakeup_cond = g_cond_new();
+  self->queue_mutex = g_mutex_new();
+
+  if (snmp_dest_counter == 0)
+    {
+      /* initialize the SNMP only once */
+      init_snmp(s_snmp_name);
+      SOCK_STARTUP;
+    }
+  ++snmp_dest_counter;
+
+  /* default values */
+  self->version = g_strdup(s_v2c);
+  self->port = 162;
+  self->community = g_strdup("public");
+  self->auth_algorithm = g_strdup(s_sha);
+  self->enc_algorithm = g_strdup(s_aes);
+  self->transport = g_strdup("UDP");
+
+  log_template_options_defaults(&self->template_options);
+  self->worker_options.is_output_thread = TRUE;
+
+  return (LogDriver *)self;
+}
+
+extern CfgParser snmpdest_dd_parser;
+
+static Plugin snmpdest_plugins[] =
+{
+  {
+    .type = LL_CONTEXT_DESTINATION,
+    .name = "snmp",
```

Comment:
Please extract these kind of objects into snmp-plugin.c, similarly to the other drivers.

Reply:
extracted